libs/parallelism/execution/include/hpx/execution/algorithms/split.hpp

```diff
@@ -0,0 +1,500 @@
+                                std::forward<Receiver>(receiver)},
+                            error);
+                    }
+
+                    void operator()(value_type const& ts)
+                    {
+                        std::visit(
+                            value_visitor<Receiver>{
+                                std::forward<Receiver>(receiver)},
+                            ts);
+                    }
+                };
+
+                void set_predecessor_done()
+                {
+                    predecessor_done = true;
+
+                    {
+                        // We require taking the lock here to synchronize with
+                        // threads attempting to add continuations to the vector
+                        // of continuations. However, it is enough to take it
+                        // once and release it immediately.
+                        //
+                        // Without the lock we may not see writes to the vector.
+                        // With the lock threads attempting to add continuations
+                        // will either:
+                        // - See predecessor_done = true in which case they will
+                        //   call the continuation directly without adding it to
+                        //   the vector of continuations. Accessing the vector
+                        //   below without the lock is safe in this case because
+                        //   the vector is not modified.
+                        // - See predecessor_done = false and proceed to take
+                        //   the lock. If they see predecessor_done after taking
+                        //   the lock they can again release the lock and call
+                        //   the continuation directly. Accessing the vector
+                        //   without the lock is again safe because the vector
+                        //   is not modified.
+                        // - See predecessor_done = false and proceed to take
+                        //   the lock. If they see predecessor_done is still
+                        //   false after taking the lock, they will proceed to
+                        //   add a continuation to the vector. Since they keep
+                        //   the lock they can safely write to the vector. This
+                        //   thread will not proceed past the lock until they
+                        //   have finished writing to the vector.
+                        //
+                        // Importantly, once this thread has taken and released
+                        // this lock, threads attempting to add continuations to
+                        // the vector must see predecessor_done = true after
+                        // taking the lock in their threads and will not add
+                        // continuations to the vector.
+                        std::unique_lock<mutex_type> l{mtx};
```

Comment:
@msimberg I know that this code is not new. However, I'm still puzzled why this strange construct is needed. The `predecessor_done` is an atomic bool - shouldn't that be visible to other threads immediately?

Reply:
The lock is needed to avoid accessing the callbacks while another thread goes into this section: https://github.com/STEllAR-GROUP/hpx/blob/438db62fe90053d493e4709b09c56f192e3d342a/libs/parallelism/execution/include/hpx/execution/algorithms/split.hpp#L331-L346. The lock can earliest be released as it is right now and latest after checking if the vector of continuations is empty.

Reply:
@hkaiser do you still have concerns about this? I'm not sure how to explain it differently but if you do still have concerns it's probably easier to take it up at the next meeting.

Reply:
@hkaiser let's discuss this separately if needed.